bundlewrap/items/zfs_dataset.py

```diff
@@ -3,137 +3,154 @@
+                        }
+                },
+                # special readonly property 'mounted'
+                'mounted': self.__zfs(f'get mounted {self.name} -p -H -o value')
+            }
+        else:
+            return None
 
     def fix(self, status):
         if status.must_be_created:
-            self.__create(self.name, status.cdict)
+            self.__create()
         else:
-            for option in status.keys_to_fix:
-                self.__set_option(self.name, option, status.cdict[option])
+            for property in status.keys_to_fix:
+                if property != 'mounted':
+                    self.__set_property(property, status.cdict[property])
+            # mount after setting mountpoint property
+            if status.cdict['mounted'] != self.__zfs(f'get mounted {self.name} -p -H -o value'):
+                mount = 'mount' if status.cdict['mounted'] == 'yes' else 'unmount'
+                self.run(f'zfs {mount} {quote(self.name)}')
+
+    def cdict(self):
+        return {
+            # previously changed properties with their default values
+            **{
+                property: self.PROPERTY_DEFAULTS.get(property)
+                    for property in self.__changed_property_names
+            },
+            # item properties with their wanted values
+            **self.__item_properties,
+            # special readonly property 'mounted'
+            'mounted': 'no' if self.__item_properties.get('mountpoint') == 'none' else 'yes',
+        }
+
+    # DEPENDENCIES
 
     def get_auto_attrs(self, items):
         pool = self.name.split("/")[0]
-        pool_item = "zfs_pool:{}".format(pool)
         pool_item_found = False
+        parent_dataset = '/'.join(self.name.split('/')[0:-1])
         needs = set()
 
         for item in items:
             if item.ITEM_TYPE_NAME == "zfs_pool" and item.name == pool:
-                # Add dependency to the pool this dataset resides on.
+                # add dependency to the pool this dataset resides on
                 pool_item_found = True
-                needs.add(pool_item)
+                needs.add(f'zfs_pool:{pool}')
             elif (
                 item.ITEM_TYPE_NAME == "zfs_dataset" and
-                self.name != item.name
+                item.name == parent_dataset
             ):
-                # Find all other datasets that are parents of this
-                # dataset.
-                # XXX Could be optimized by finding the "largest"
-                # parent only.
-                if self.name.startswith(item.name + "/"):
-                    needs.add(item.id)
-                elif (
-                    self.attributes.get('mountpoint') and
-                    item.attributes.get('mountpoint') and
-                    self.attributes['mountpoint'].startswith(item.attributes['mountpoint'])
+                # add dependency to parent dataset
+                needs.add(item.id)
```

Comment:
We should also check wether the parent dataset does exist (like we already do for pool items). Otherwise the creation will fail.